### PR TITLE
Adding basepath, urlpath, hostname and ssl_policy as config

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp #6920 
Fixing #6720 

- Adding `system.ssl_policy` to config
- Adding `system.urlpath` to config
- Adding `system.url` to config
The new, generated config-file looks like:
```php
<?php

// Local configuration

// If you're unsure about what any of the config keys below do, please check the config/defaults.config.php for detailed
// documentation of their data type and behavior.

return [
	'database' => [
		'hostname' => 'localhost',
		'username' => 'friendica',
		'password' => 'friendica',
		'database' => 'friendica',
		'charset' => 'utf8mb4',
	],

	// ****************************************************************
	// The configuration below will be overruled by the admin panel.
	// Changes made below will only have an effect if the database does
	// not contain any configuration for the friendica system.
	// ****************************************************************

	'config' => [
		'php_path' => '/usr/bin/php',
		'admin_email' => '',
		'sitename' => 'Friendica Social Network',
		'hostname' => 'friendica.local',
		'register_policy' => \Friendica\Module\Register::OPEN,
		'max_import_size' => 200000,
	],
	'system' => [
		'urlpath' => '',
		'ssl_policy' => '0',
		'basepath' => '/vagrant',
		'default_timezone' => 'America/Los_Angeles',
		'language' => 'en',
	],
];
```

Added a new "Base setting" template, which looks like:
screenshot_2019-03-06 friendica social network install

It works with different sub-paths, URLs, hostnames, ...
I tested it on my local node with normal web installation and automatic installation a few times.

Nevertheless, it's not finished yet:
- [x] Adapt business logic when moving a node
- [x] Fixing the unit tests (and enhance them)
- [x] Migration logic (moving the basepath, urlpath, hostname out of the DB into the config)

<strike>After merging this PR, I'll start refactoring the `$basepath` usage, because we don't need it anymore inside `App` (it get's loaded only out of the config and not more with the DB)</strike> done

### CODE REVIEW

**Comment by MrPetovan** [comment]
Why did you include `ssl_policy` in this move? Is it guessed on every request like `urlpath` and `basepath` were?

**Comment by nupplaphil** [comment]
Because I started first with creating the full `url` and write it to the config. After a while, I recognized that the `baseurl` is depending on the `ssl` flag at each call, so I reverted the base url creating/saving during the installation process, because we cannot save business logic at this part.

Nevertheless I found that the SSL policy is still an important setting for the base of the Friendica node, so I thought it would be fine letting it as is.

File: src/Util/Network.php [hunk 1]
```diff
@@ -836,4 +836,29 @@ public static function unparseURL($parsed)
...
+	/**
+	 * @brief Switch the scheme of an url between http and https
+	 *
+	 * @param string $url URL
+	 *
+	 * @return string switched URL
+	 */
+	public static function switchScheme($url)
+	{
+		$parts = parse_url($url);
```
Comment (MrPetovan) [inline src/Util/Network.php]: Please use the `PHP_URL_SCHEME` constant to limit the output of `parse_url()` to the only relevant bit.

**Comment by nupplaphil** [comment]
Tested:
- Installation wizard (with failures, with subppath, with different SSL-settings)
- Automatic Installation (including new tests)
- Migration path (1306 => 1308)
- Admin settings (new SSL settings -> None, full, self-signed)
- Relocation process

I hope that's enough ..

File: config/dbstructure.config.php [hunk 2]
```diff
@@ -34,7 +34,7 @@
 use Friendica\Database\DBA;
 
 if (!defined('DB_UPDATE_VERSION')) {
-	define('DB_UPDATE_VERSION', 1307);
+	define('DB_UPDATE_VERSION', 1308);
```
Comment (MrPetovan) [inline config/dbstructure.config.php]: This should be bumped to `1309`, I'm surprised you didn't get a conflict after #6939 was merged.

File: mod/admin.php [hunk 3]
```diff
@@ -1093,9 +1093,10 @@ function update_table(App $a, $table_name, $fields, $old_url, $new_url)
 		// update config
 		$configFileSaver = new \Friendica\Util\Config\ConfigFileSaver($a->getBasePath());
 		$configFileSaver->addConfigValue('config', 'hostname', parse_url($new_url, PHP_URL_HOST));
+		$configFileSaver->addConfigValue('system', 'urlpath', parse_url($new_url, PHP_URL_PATH));
+		$configFileSaver->addConfigValue('system', 'url', $new_url);
```
Comment (MrPetovan) [inline mod/admin.php]: Is this comma-safe?

Comment (nupplaphil) [reply mod/admin.php]: What do you mean? The `$new_url` is already used several times before for updates

Comment (nupplaphil) [reply mod/admin.php]: Ah, are you talking about the comma bug of the configsaver? No, not yet.

In the meantime, there are two bugs on my radar:
- no-comma at the end
- missing category will lead to skip adding

**Comment by MrPetovan** [comment]
And **now** there's a conflict.

**Comment by MrPetovan** [comment]
Still conflicts, unfortunately.

**Comment by nupplaphil** [comment]
I'll temporary close it, because I'm currently fixing #6941 first.

**Pull request closed without merging**